### PR TITLE
How many calorie (CanOPEN edition)

### PR DESCRIPTION
**Pull request by oleglazari**
Added a python script for battery calorimetry testing. The script does the following

- Asks the user running the script for a specific COM port.
- Reads from the COM port into a new CSV file
- Logs The Temperature Array and the Error Array into the CSV File

Upgraded to the newly written CanOPEN stack. This simplifies most of the Can stuff

- Refactored the usage of data (uint32 --> uint8s and uint16s).
- Added new methods in TMU.cpp and TMU.hpp respectively
- Refactored DEV1-TMU target as well!

### CODE REVIEW

**Comment by oleglazari** [comment]
[Battery Calorimetry Test Support](https://trello.com/c/L8SlRzKi/189-battery-calorimetry-test-support)

File: CMakeLists.txt [hunk 1]
```diff
@@ -41,7 +41,9 @@ project(${BOARD_LIB_NAME}
         LANGUAGES CXX C
         )
 
-add_library(${PROJECT_NAME} STATIC)
+add_library(${PROJECT_NAME} STATIC
```
Comment (mjmagee991) [inline CMakeLists.txt]: Revert this change. CLion thinks it's smart, but it isn't

File: targets/cal_test/cal_to_csv.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import serial
+import csv
+import time
+
+# Open the serial port
+com_port = input('Enter the COM port (Format: COMXX): ')
```
Comment (mjmagee991) [inline targets/cal_test/cal_to_csv.py]: Technically, it's only formatted as `COMXX` on Windows. It's named differently on Mac and Linux, but those users should be able to figure it out for themselves

File: targets/cal_test/cal_to_csv.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ser = serial.Serial(com_port, 9600)  # Replace the COM port with the one you're using
+
+try:
+    while True:
```
Comment (mjmagee991) [inline targets/cal_test/cal_to_csv.py]: This while loop is going to cause the CSV file to be reopened and overwritten if a KeyboardInterrupt occurs. It doesn't look like you really need this here, but if you think you do, you can add something that checks if the file already exists and appends or starts a new file, so you don't accidentally delete all the data

File: targets/cal_test/main.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
+#include <EVT/io/UART.hpp>
```
Comment (mjmagee991) [inline targets/cal_test/main.cpp:6]: Add a comment at the top explaining the purpose of the target

File: targets/cal_test/cal_to_csv.py
Comment (mjmagee991) [inline targets/cal_test/cal_to_csv.py:1]: Move this file to a new top-level folder called tools

File: targets/cal_test/main.cpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    while (true) {
+        //print temps and error array for csv friendly format
+        //temp1: ,temp2: ,temp3: ,temp4:, error1: ,error2: ,error3: ,error4:
+        uint16_t temps[4];
+        TMU::DEV::MAX31855::MaxStatus errors[4];
+        for (int i = 0; i < 4; i++) {
+            errors[i] = MAXES[i].readTemp(temps[i]);
+        }
+        uart.printf("temp1:%d, temp2:%d, temp3:%d, temp4:%d, error1:%d, error2:%d, error3:%d, error4:%d\r\n", temps[0], temps[1], temps[2], temps[3], errors[0], errors[1], errors[2], errors[3]);
```
Comment (mjmagee991) [inline targets/cal_test/main.cpp]: For CSV files, we don't want the header in every line. We want `temp1, temp2, temp3, ...` once at the beginning of the file and then we just want the data, like `51, 42, 33, 41, ...`

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
A couple minor changes. Also, make sure you back merge master before you try to merge because you have some changes missing

**Comment by oleglazari** [comment]
build is actively failing, just a skill issue I guess when back merging will fix.

**Comment by mjmagee991** [comment]
When you merged your CANopen upgrade branch into this branch, you brought over the updated submodule, which is breaking everything. Why did you merge that branch here?

**Comment by oleglazari** [comment]
> When you merged your CANopen upgrade branch into this branch, you brought over the updated submodule, which is breaking everything. Why did you merge that branch here?

lowkey kinda misread when u said back merge and thought submodule update ¯\_(ツ)_/¯

File: targets/DEV1-TMU/main.cpp [hunk 6]
```diff
@@ -64,10 +62,6 @@ extern "C" void CONodeFatalError(void) {}
 
 extern "C" void COIfCanReceive(CO_IF_FRM* frm) {}
```
Comment (ActuallyTaylor) [inline targets/DEV1-TMU/main.cpp]: You should remove all of these these externs

File: targets/cal_test/cal_to_csv.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        # Generate a new file name with timestamp
+        csv_file_path = f'sensor_data.csv'
```
Comment (ActuallyTaylor) [inline targets/cal_test/cal_to_csv.py]: You don't need an fstring here

**Review by ActuallyTaylor — CHANGES_REQUESTED** [review]

**Comment by oleglazari** [comment]
[Upgrade TMU CANopen](https://trello.com/c/uQv4J4no/228-upgrade-tmu-canopen)

File: CMakeLists.txt
Comment (mjmagee991) [inline CMakeLists.txt:1]: Revert your changes to this file

File: targets/DEV1-TMU/main.cpp [hunk 8]
```diff
@@ -1,19 +1,12 @@
...
-#include <EVT/dev/Timer.hpp>
+#include <EVT/io/CANOpenMacros.hpp>
 #include <EVT/io/CANopen.hpp>
 #include <EVT/io/UART.hpp>
 #include <EVT/io/pin.hpp>
 #include <EVT/manager.hpp>
 #include <EVT/utils/log.hpp>
 #include <EVT/utils/time.hpp>
 
+#include "../../libs/EVT-core/samples/canopen/canopen_tpdo/TPDOCanNode.hpp"
```
Comment (mjmagee991) [inline targets/DEV1-TMU/main.cpp]: Fix this include

File: targets/DEV1-TMU/main.cpp [hunk 9]
```diff
@@ -39,64 +32,42 @@ namespace DEV = EVT::core::DEV;
  * @param message[in] The passed in CAN message that was read.
  */
 
+IO::UART& uart = IO::getUART<IO::Pin::UART_TX, IO::Pin::UART_RX>(9600);
```
Comment (mjmagee991) [inline targets/DEV1-TMU/main.cpp]: Don't have a global uart

File: targets/DEV1-TMU/main.cpp [hunk 10]
```diff
@@ -39,64 +32,42 @@ namespace DEV = EVT::core::DEV;
...
-extern "C" void COPdoTransmit(CO_IF_FRM* frm) {}
-
-extern "C" int16_t COPdoReceive(CO_IF_FRM* frm) { return 0; }
-
-extern "C" void COPdoSyncUpdate(CO_RPDO* pdo) {}
-
-extern "C" void COTmrLock(void) {}
-
-extern "C" void COTmrUnlock(void) {}
+//TPDO event handler to print the raw TPDO message when sending
```
Comment (mjmagee991) [inline targets/DEV1-TMU/main.cpp]: Remove this

File: targets/DEV1-TMU/main.cpp [hunk 11]
```diff
@@ -121,85 +92,69 @@ int main() {
         TMU::DEV::MAX31855(spi, 3),
     };
 
+    //tpdo node
```
Comment (mjmagee991) [inline targets/DEV1-TMU/main.cpp]:
```suggestion
```

**Review by mjmagee991 — CHANGES_REQUESTED** [review]

File: targets/DEV1-TMU/main.cpp [hunk 12]
```diff
@@ -1,20 +1,11 @@
-/**
```
Comment (mjmagee991) [inline targets/DEV1-TMU/main.cpp:1]: Add this target explanation back in

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
Some things still haven't been resolved

**Review by mjmagee991 — APPROVED** [review]
Very epic and cool

**Review by ActuallyTaylor — APPROVED** [review]

**Pull request merged**